turbo/app/reset-datadir.go

```diff
@@ -0,0 +1,216 @@
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	g "github.com/anacrolix/generics"
+	"github.com/anacrolix/torrent/metainfo"
+	"github.com/erigontech/erigon-db/rawdb"
+	"github.com/erigontech/erigon-lib/chain"
+	"github.com/erigontech/erigon-lib/chain/snapcfg"
+	"github.com/erigontech/erigon-lib/common/datadir"
+	"github.com/erigontech/erigon-lib/kv"
+	"github.com/erigontech/erigon-lib/log/v3"
+	"github.com/erigontech/erigon/cmd/utils"
+	"github.com/erigontech/erigon/core"
+	"github.com/erigontech/erigon/node"
+	"github.com/erigontech/erigon/turbo/debug"
+	"github.com/urfave/cli/v2"
+)
+
+var (
+	pruneFlag = cli.BoolFlag{
+		Name:     "prune",
+		Usage:    "Remove files not described in snapshot set.",
+		Value:    false,
+		Aliases:  []string{"p"},
+		Category: "Reset",
+	}
+	dryRunFlag = cli.BoolFlag{
+		Name:     "dry-run",
+		Usage:    "Print files that would be removed, but do not remove them.",
+		Value:    false,
+		Aliases:  []string{"n"},
+		Category: "Reset",
+	}
+)
+
+func resetCliAction(cliCtx *cli.Context) (err error) {
+	// Set logging verbosity. Oof that function signature.
+	logger, _, _, _, err := debug.Setup(cliCtx, true)
+	if err != nil {
+		err = fmt.Errorf("setting up logging: %w", err)
+		return
+	}
+	prune := pruneFlag.Get(cliCtx)
+	dryRun := dryRunFlag.Get(cliCtx)
+	dataDirPath := cliCtx.String(utils.DataDirFlag.Name)
+
+	configChainName, err := getChainNameFromChainData(cliCtx, logger)
+	if err != nil {
+		return fmt.Errorf("getting chain name from chaindata: %w", err)
+	}
+	chain := utils.ChainFlag.Get(cliCtx)
+	if cliCtx.IsSet(utils.ChainFlag.Name) {
+		if configChainName.Ok && configChainName.Value != chain {
+			// Pedantic but interesting.
+			logger.Warn("chain name flag and chain config do not match", "flag", chain, "config", configChainName.Value)
+		}
+		logger.Info("using chain name from flag", "chain", chain)
+	} else {
+		if !configChainName.Ok {
+			return errors.New("chain flag not set and chain name not found in chaindata (reset already occurred or invalid data dir?)")
+		}
+		chain = configChainName.Unwrap()
+		logger.Info("read chain name from config", "chain", chain)
+	}
+
+	dirs := datadir.Open(dataDirPath)
+	unlock, err := dirs.TryFlock()
+	if err != nil {
+		return fmt.Errorf("failed to lock data dir %v: %w", dirs.DataDir, err)
+	}
+	defer unlock()
+	err = snapcfg.LoadRemotePreverified(cliCtx.Context)
+	if err != nil {
+		// TODO: Check if we should continue? What if we ask for a git revision and
+		// can't get it? What about a branch? Can we reset to the embedded snapshot hashes?
+		return fmt.Errorf("loading remote preverified snapshots: %w", err)
+	}
+	cfg := snapcfg.KnownCfg(chain)
+	// Should we check cfg.Local? We could be resetting to the preverified.toml...?
+	logger.Info(
+		"Loaded preverified snapshots hashes",
+		"len", len(cfg.Preverified.Items),
+		"chain", chain,
+	)
+	removeFunc := func(path string) error {
+		return os.Remove(filepath.Join(dirs.Snap, path))
```

Comment:
also need remove `.torrent`?

Reply:
This does inside the walk function. It removes torrent files if their hashes don't match, and it removes any file if "prune" is on, and they aren't listed in the snapshot hashes (excluding of course the associated torrent file for a given item name).